### ISSUE TITLE
How to use with autocommit=false

### ISSUE THREAD

**Issue by tomhillgreyridge** [kagkarlsson/db-scheduler#202]
I know there have been several bug reports / fixes / rollbacks of fixes to do with how db-scheduler works with the autocommit property on jdbc connections.  As far as I can tell, it still doesn't appear to work correctly, so I thought I would provide an explicit explanation of what I would expect to happen and why and some example code demonstrating why I don't think it currently works.

In most applications I have worked on, there are two types of jobs

- Background jobs which occur on a repeating schedule (nightly batch processing, end of the month runs, etc)
- Jobs which are submitted based on user actions i.e. user creates an invoice which triggers a job to reconcile accounts

99% of applications I have worked with have autocommit set to false for the JDBC connection in the connection configuration to force correct transaction handling (autocommit=true is almost always a bug).  Effectively, we want to be able to rollback the submission of a job if later code fails due to an error (i.e. user creates an invoice, this triggers the job but then the auditing of the creation fails - we want the invoice to be rolled back *and* the submission of the job to be rolled back).

The problem is, in Spring Boot, when I have autocommit=false in my configuration file I cannot find a way to get db-scheduler to create the recurring tasks as it does not commit them!  So, I started off with an application.yml which looks like

```
spring:
  datasource:
    url: "<url goes here>"
    username: "<username>"
    password: "<password>"
    hikari:
      autocommit: false
```

and tried to define tasks as demonstrated in the spring boot example i.e.

```
@Configuration
class JobConfiguration {
  @Bean
  fun recurringTask(): Task<Void>{
    return Tasks
      .recurring("mytask", Schedules.cron("*/5 * * ? * *"))
      .execute { taskInstance, executionContext ->
        println("Job has been called")
      }
  }
}
```

The logging shows the task being picked up but, since the connection is autocommit false, they never get committed (table stays empty) so the job never runs.  I figured I could get round this by skipping the spring boot autoconfiguration, registering the tasks manually and then committing the datasource, so I tried the following

```
@Configuration
class JobConfiguration {

  @Bean
  fun scheduler(datasource: DataSource): Scheduler {
    val builder = SchedulerBuilder(datasource, emptyList())

    builder.startTasks(
      Tasks
        .recurring("testtask",Schedules.fixedDelay(Duration.ofSeconds(5)))
        .execute { taskInstance, executionContext ->  println("This is a task")}
    )

    //try committing after the tasks are defined
    datasource.connection.commit()

    val scheduler = builder.build()

    scheduler.start()

    //try committing after the scheduler is started since this seems to be when the tasks are inserted
    datasource.connection.commit()
    
    return scheduler
  }
}
```
but that doesn't seem to work either (I can see my code being executed but the table stays empty).  I was hoping that by committing the same datasource which is used to create the scheduler that would commit the insert of the scheduled tasks, but it doesn't seem to.

In terms of how I would expect it to work, the following would seem logical to me

- The recurring tasks to be committed automatically since they are clearly not in a transaction (they are created on system startup as a one-off action which must succeed or fail rather than in response to use actions - as such I would say there is an expectation from the developer that these get committed automatically).
- The one off tasks would use the underlying connection so that the application can choose whether to commit or rollback

I'd be equally happy if I could find a way to *manually* commit the recurring tasks.  I could obviously change autocommit to true but that means, if someone forgets to use @Transactional on a piece of code, we have code in the application which is non-transactionally consistent, which I'd rather avoid.

Is there any chance you could add a demonstration of how to get db-scheduler working with spring boot where autocommit is false?

**Comment by kagkarlsson** [comment]
I agree that Scheduler-internal operations should always commit, disregarding autocommit setting on the connection. This is why I created [two separate JdbcTaskRepositories](https://github.com/kagkarlsson/db-scheduler/blob/master/db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/SchedulerBuilder.java#L193), one for internal operations (always commit) and one for client-operations (if autocommit=false, commit based on setting)

What you are describing sounds like a bug, since the scheduling of recurring tasks is not client-initiated as you say. From what I can see, it looks like the [onStartup items](https://github.com/kagkarlsson/db-scheduler/blob/master/db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/Scheduler.java#L121) are executed using the `JdbcTaskRepository` for the client, but it should be using the one for the internal scheduler-operations which will always commit.

**Comment by kagkarlsson** [comment]
Hopefully addressed in #204

**Comment by kagkarlsson** [comment]
Should be fixed in 10.3. Please let me know if that does not fix your problem
